src/main/java/org/springframework/integration/kafka/core/DefaultConnectionFactory.java

```diff
@@ -172,12 +103,17 @@ public void setFetchMetadataTimeout(int fetchMetadataTimeout) {
 	 */
 	@Override
 	public BrokerAddress getLeader(Partition partition) {
-		try {
-			lock.readLock().lock();
```

Comment:
Yes, but `ReentrantReadWriteLock` does not support write upgrades. 

What do you mean with the comment about `getMetadataCache()` ?